### PR TITLE
fix: Android - make sure logcat.log is attached to unhandled event (SIGSEGV Segfault)

### LINKED ISSUE #3461: logcat.log is not attached to unhandled event (SIGSEGV Segfault)
### Package

Sentry.Maui

### .NET Flavor

.NET

### .NET Version

8.0.6

### OS

Android

### SDK Version

4.7.0

### Self-Hosted Sentry Version

_No response_

### Steps to Reproduce

```
options.Android.LogCatIntegration = LogCatIntegrationType.Unhandled;
```

I don't know how to provoke `SIGSEGV Segfault`.

### Expected Result

It is expected that the error event has an attachment `logcat.log`.

### Actual Result

There are no attachments to the event.

### PR DESCRIPTION
**Pull request by bricefriha**
From [sentry-native#1027](https://github.com/getsentry/sentry-native/pull/1027), we had to add the handler strategy to the .NET SDK, from the sentry-java ([7.15.0-alpha1](https://github.com/getsentry/sentry-java/releases/tag/7.15.0-alpha.1)).

This is also meant to fix #3461. I tested this, and it seems to be fixed now.

More context at:
* https://github.com/dotnet/android/issues/9055#issuecomment-2261347912

### CODE REVIEW

File: src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj
Comment (bruno-garcia) [inline src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:1]: some BOM or setting in your IDE is adding these to the diff, please address that

File: src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj
Comment (bruno-garcia) [inline src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:40]: I don't understand what's changing here, is this some IDE messing up?

File: src/Sentry/NdkHandlerStrategy.cs
Comment (bruno-garcia) [inline src/Sentry/NdkHandlerStrategy.cs:13]: This is Android specific, but it was added to the whole SDK regardless of platform. Please add under `/Android/`

It should only show up if you're `#if ANDROID`

File: src/Sentry/SentryOptions.cs
Comment (bruno-garcia) [inline src/Sentry/SentryOptions.cs:88]: nit
```suggestion
```

File: src/Sentry/SentryOptions.cs
Comment (bruno-garcia) [inline src/Sentry/SentryOptions.cs:82]: Please describe further what this is for.
Consider adding `<seealso href="https://github.com/dotnet/android/issues/9055#issuecomment-2261347912">`

File: src/Sentry/SentryOptions.cs
Comment (bruno-garcia) [inline src/Sentry/SentryOptions.cs:80]: Why do we need the private backing field?

Couldn't this be a 1 liner?

```csharp
public NdkHandlerStrategy HandlerStrategy {get;set;} = NdkHandlerStrategy.SENTRY_HANDLER_STRATEGY_DEFAULT;
```

File: test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs
Comment (bruno-garcia) [inline test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs:59]: nit:
we add lines between test
```suggestion

    [Fact]
```

File: test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs
Comment (bruno-garcia) [inline test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs:73]: we're just testing setters here, can we add some integration test somehow?

Can we repro this null reference to SIGSEGV issue? We have tests in the repo where we have the SDK capture something and we grab the event from the transport and asset on what was created.

We'd add tests and `#if ANDROID`. cc @jamescrosswell @vaind are DeviceTests alive? I know we had issues
If so, any points you could give @bricefriha on how to do this?

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bruno-garcia) [inline src/Sentry/Platforms/Android/SentrySdk.cs:63]: nit: inconsistent. You're capitalizing Java but not .NET
Either go with Java and .NET or, java and .net
```suggestion
        // Convert the .NET NdkHandlerStrategy to Java / Android
```

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bruno-garcia) [inline src/Sentry/Platforms/Android/SentrySdk.cs:65]: I'm struggling to parse. Why would `Values()` return null?
Can we use `var`
```suggestion
        var handlerStrategy = JavaSdk.Android.Core.NdkHandlerStrategy.Values()[(int)options.HandlerStrategy];
```

An extension method on `NdkHandlerStrategy` is probably a good solution:

```csharp
ToJava()
```

File: src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj
Comment (bricefriha) [reply src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:40]: yes it's VS that updates the indentation. I should live this out of git staging next time

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bricefriha) [reply src/Sentry/Platforms/Android/SentrySdk.cs:65]: I'll try the `ToJava()` thanks for the suggestion

File: src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj
Comment (bricefriha) [reply src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:6]: it was more so for the sake of testing locally.
e.g. if you pull the branch and want to run it, you just have to rebuild the android bindings and you should be good to go

File: src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj
Comment (bricefriha) [reply src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:1]: i'm just gonna be more meticulous with git staging next time

File: src/Sentry/SentryOptions.cs
Comment (bricefriha) [reply src/Sentry/SentryOptions.cs:80]: absolutely, I don't know why I did it this way. 
I think I was going for another approach at first

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bricefriha) [reply src/Sentry/Platforms/Android/SentrySdk.cs:65]: > Why would Values() return null?

@bruno-garcia In my experience, it's always better to assume that a method can return null, even if it seems impossible. I quite feel strongly about it. 
We would rather get it to default to `SENTRY_HANDLER_STRATEGY_DEFAULT` than throw an unhandled exception which would cause things to break somewhere else that would be harder to troubleshoot in 10 months. `Values()?` doesn't have any cost compared to `Values()`. 

Now `JavaSdk.Android.Core.NdkHandlerStrategy?` is silly, let's do a JavaSdk.Android.Core.NdkHandlerStrategy.Values()?[handlerStrat] ?? JavaSdk.Android.Core.NdkHandlerStrategy.SENTRY_HANDLER_STRATEGY_DEFAULT

File: src/Sentry/NdkHandlerStrategy.cs
Comment (bruno-garcia) [inline src/Sentry/NdkHandlerStrategy.cs:25]: This means we end up with an empty class on platforms other than Android. Why would we do that?

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bruno-garcia) [inline src/Sentry/Platforms/Android/SentrySdk.cs:93]: Why do we need the null coalesce operator if `ToJava` doesn't return null?

Besides complexity an additional issue with this is that now we have 2 places dealing with null and defining a default. Had we even needed to change the default we'd need to 'remember' to change these two spots

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bruno-garcia) [inline src/Sentry/Platforms/Android/SentrySdk.cs:192]: This change doesn't seem related to the goal of this PR. Or is it?

Why was `configuration` arg removed?

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bruno-garcia) [inline src/Sentry/Platforms/Android/SentrySdk.cs:196]: Could you clarify why we need this cast now?

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bricefriha) [reply src/Sentry/Platforms/Android/SentrySdk.cs:192]: mistake my bad

File: src/Sentry/Platforms/Android/SentrySdk.cs
Comment (bruno-garcia) [reply src/Sentry/Platforms/Android/SentrySdk.cs]: > We would rather get it to default to SENTRY_HANDLER_STRATEGY_DEFAULT than throw an unhandled exception which would cause things to break somewhere else that would be harder to troubleshoot in 10 months. Values()? doesn't have any cost compared to Values().

It's more nuanced than that. With the `??` approach we also will scratch our heads when things are not working the way we expected: "I set the handler to A, why is B happening"?

You should add in code or comments that intention you describe in this comment then. Why would it be null in this case? Is some possible bindings problem? Is it the caller could be on a version of C# pre-nullability?
Once you clarify why that could be the case, you can use our debug logger. Try to read the value and if it's null you use `LogError("expected a value set for handler since a default is hard coded but got null")

IMO that's overkill, and throwing `??` all over the code base "just in case" is far from ideal.

That said I share the sentiment that we can't foresee all issues and this is a crash reporting SDK so we need ot be careful. The way we go about that is by using error handling and our debug logging. You could do that on `ToJava` only, and avoid the other `??` usage. `ToJava` would require the internal `ILogger` as an argument to properly log out the error. But unless there's a reasonable suspicion (based on experience or otherwise) that reading that value could be null even though we're hard coding values all over the place, it's enough to have a higher level try/catch deal (debug logging, always) with it and avoid the app from crashing.

Comment (bricefriha) [reply src/Sentry/Platforms/Android/SentrySdk.cs]: fair enough, you're right

File: src/Sentry/Platforms/Android/NdkHandlerStrategy.cs
Comment (jamescrosswell) [inline src/Sentry/Platforms/Android/NdkHandlerStrategy.cs:2]: We prefer File-scoped namespace declaration over Block-scoped namespace declaration in this repo.

File: test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs
Comment (jamescrosswell) [reply test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs:74]: > cc @jamescrosswell @vaind are DeviceTests alive? I know we had issues

They're working fine on the main branch. We've got issues with them when we try to make them target net8.0-android or net8.0-ios... we don't have to do that until we bump to net9.0 though.

File: src/Sentry/SentryOptions.cs
Comment (bricefriha) [reply src/Sentry/SentryOptions.cs:93]: Yes, now that you mention it, I think it would make more sense to have `SENTRY_HANDLER_STRATEGY_CHAIN_AT_START ` as a default as a constant, as I don't see any scenario where we wouldn't use it. Especially since the issue comes from dotnet-generated machine code

**Comment by bricefriha** [comment]
@jamescrosswell it seems like the wrong version of the Java SDK is set there. 
Which causes the CI to crash.  
image

Is there a way to set this version to 7.15.0-alpha.1 without updating Sentry.Bindings.Android.csproj?

**Comment by jamescrosswell** [comment]
> Is there a way to set this version to `7.15.0-alpha.1` without updating `Sentry.Bindings.Android.csproj`

I think we do want to update the Sentry.Bindings.Android.csproj right? How would we be using the new Native functionality otherwise?

File: src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj
Comment (bricefriha) [inline src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:6]: We changed the SdkVersion to `7.15.0-alpha.1` just for the alpha-release and to make it easy for CI. We can remove it just before merging to `main`

[more](https://github.com/getsentry/sentry-dotnet/pull/3694#issuecomment-2430134945)

Comment (jamescrosswell) [reply src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:6]: OK, is the plan to wait for this to be shipped in the main branches of the native and android libraries first then? Have we confirmed this resolves the original issue that it was meant to address?

Comment (jamescrosswell) [reply src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:6]: OK cool. So maybe we don't merge this into main just yet... we can make an alpha release from this branch "so folks can try that out".

Thanks heaps @bricefriha - nice work!

Comment (bruno-garcia) [reply src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:6]: with this we're markign only this package as alpha. How do we release from the branch? If I set the whole repo's version `-alpha`, this package will be `-alpha0.1-alpha0.1`.

Comment (jamescrosswell) [reply src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:6]: @bruno-garcia I thought this indicated we depend on an alpha of the Sentry Android SDK (it doesn't mark any of the packages in the sentry-dotnet repo as alpha). Is that not the case?

Comment (bruno-garcia) [reply src/Sentry.Bindings.Android/Sentry.Bindings.Android.csproj:6]: lets make an alpha of this branch alone (the whole SDK can be alpha).
otherwise we need to merge into `main` and if we make previews from there, we'll have teh issue I mentioend

File: test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs
Comment (bruno-garcia) [inline test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs:64]: Why do we have logcat tests on the NDK PR?

Comment (bricefriha) [reply test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs:64]: As mentioned in the description, this PR is mostly to fix #3461. The issue was that the logcat file was not attached.

So, it's crucial to test that the logcat file is being attached properly

Comment (jamescrosswell) [reply test/Sentry.Maui.Tests/SentryMauiOptionsTests.cs:64]: @bricefriha I think [that comment](https://github.com/dotnet/android/issues/9055#issuecomment-2261347912) describes a specific situation in which logcat logs aren't sent. More generally, as long as we're not dealing with SEGSEGV exceptions like that, we should be getting logcat reports come through though, so it should be possible to test this fix without mixing the two PRs right?

File: CHANGELOG.md
Comment (jamescrosswell) [inline CHANGELOG.md:8]:
```suggestion
```

This isn't an option right? We always use `SENTRY_HANDLER_STRATEGY_CHAIN_AT_START` when initialising the integration from the .NET SDK?

File: CHANGELOG.md
Comment (jamescrosswell) [inline CHANGELOG.md:9]: Isn't this a comment from another PR?

File: CHANGELOG.md
Comment (jamescrosswell) [inline CHANGELOG.md:7]: Is that this PR or another PR? The title talks about attachments but the changelog talks about fixing a `NullReferenceException`...

File: CHANGELOG.md
Comment (bricefriha) [reply CHANGELOG.md:8]: It isn't anymore indeed, I forgot to update it. thanks for noticing this

File: CHANGELOG.md
Comment (bricefriha) [reply CHANGELOG.md:8]: Not an open one at least. 
this is on [main](https://github.com/getsentry/sentry-dotnet/blob/main/CHANGELOG.md#L12) for sure

File: CHANGELOG.md
Comment (bricefriha) [reply CHANGELOG.md:7]: it's not from here, I have no idea how it ended up here

Comment (bricefriha) [reply CHANGELOG.md]: similarly it's coming from me

either way it's fixed now

File: CHANGELOG.md
Comment (jamescrosswell) [reply CHANGELOG.md:8]: Ah I see, it's just been moved from below the `###Features` in this branch. I was wondering why it was showing as being added in this PR.

Comment (bricefriha) [reply CHANGELOG.md:8]: yeah I see

do you think I should change that?

Comment (jamescrosswell) [reply CHANGELOG.md:8]: Historically we've put the features before the Fixes... just a question of what we want to draw people's attention to first. Not a big deal.

**Comment by jamescrosswell** [comment]
Requested [this release](https://github.com/getsentry/publish/issues/4545) that we can share with testers.

**Comment by bruno-garcia** [comment]
Note: https://github.com/getsentry/publish/issues/4545#issuecomment-2452007323

**Comment by bruno-garcia** [comment]
which issue to reach out for feedback? anyone trying this out to help validate?

**Comment by jamescrosswell** [comment]
> which issue to reach out for feedback? anyone trying this out to help validate?

See comment:
- https://github.com/getsentry/sentry-dotnet/issues/3461#issuecomment-2453766597

**Comment by bricefriha** [comment]
> anyone trying this out to help validate?

@bruno-garcia @jamescrosswell  I created a repo to test this alpha release and it works perfectly https://github.com/bricefriha/Sentry-3694

**Comment by bruno-garcia** [comment]
> > anyone trying this out to help validate?
> 
> @bruno-garcia @jamescrosswell I created a repo to test this alpha release and it works perfectly [bricefriha/Sentry-3694](https://github.com/bricefriha/Sentry-3694)

Awesome! thank you for that.
Shoudl we merge?

**Comment by bricefriha** [comment]
>Shoudl we merge?

We could, but the issue is that we need to use the Android SDK version 7.15.0-alpha.1 until the NDKHandler setting is available on the full release. So shouldn't we wait for that first?

**Comment by bruno-garcia** [comment]
> > Shoudl we merge?
> 
> We could, but the issue is that we need to use the Android SDK version 7.15.0-alpha.1 until the NDKHandler setting is available on the full release. So shouldn't we wait for that first?

If it's validated, we should reach out to folks on Android to merge that to `main` and ship a new release.

We can merge ours (after bumping to that version) then.

Also: Should that option be the new default? So users don't need to do anything?

**Comment by bricefriha** [comment]
>We can merge ours (after bumping to that version) then.

sounds good 

>Should that option be the new default? So users don't need to do anything?

yes it is, so basically users don't need to change anything at all

**Comment by bruno-garcia** [comment]
What's the next steps for this?

**Comment by bricefriha** [comment]
> What's the next steps for this?

I let the Java team know that we are ready to merge so they are working on merging their PRs as well

once they have the update on the main release, we can merge

**Comment by bruno-garcia** [comment]
merged! https://github.com/getsentry/sentry-native/pull/1027#event-15314896031

We'd like to wait for their release though, so we can checkout the release tag instead of a random commit sha

**Comment by bricefriha** [comment]
> We'd like to wait for their release though

The release is out now!
https://github.com/getsentry/sentry-native/releases/tag/0.7.13

We just need an update from `sentry-java`

**Comment by jamescrosswell** [comment]
> The release is out now! https://github.com/getsentry/sentry-native/releases/tag/0.7.13
> 
> We just need an update from `sentry-java`

@markushi any timeframe on getting the fix we tested in [7.15.0-alpha.1](https://github.com/getsentry/sentry-java/releases/tag/7.15.0-alpha.1) into an official release? I'd love to roll this out in the .NET SDK but we've got a bit of a dependency chain.

Looks like we'll have to do something similar for iOS as well, and it would be nice to reference this solution when discussing with the cocoa team.

**Comment by markushi** [comment]
> > The release is out now! [getsentry/sentry-native@`0.7.13` (release)](https://github.com/getsentry/sentry-native/releases/tag/0.7.13)
> > We just need an update from `sentry-java`
> 
> @markushi any timeframe on getting the fix we tested in [7.15.0-alpha.1](https://github.com/getsentry/sentry-java/releases/tag/7.15.0-alpha.1) into an official release? I'd love to roll this out in the .NET SDK but we've got a bit of a dependency chain.
> 
> Looks like we'll have to do something similar for iOS as well, and it would be nice to reference this solution when discussing with the cocoa team.

If it's tested and we know the change works, we can merge and get this released at any time. Let me get the [PR](https://github.com/getsentry/sentry-java/pull/3671) up-to-date first.

**Comment by jamescrosswell** [comment]
Looks like the fix was included with the [7.19.0 release of the Java SDK](https://github.com/getsentry/sentry-java/blob/main/CHANGELOG.md#7190). We just need to bump the version in this PR and we should be good to go!

**Review by jamescrosswell — APPROVED** [review]

**Pull request merged**
